fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/synchronizedsession/SynchronizedSession.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * SynchronizedSession.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb.synchronizedsession;
+
+import com.apple.foundationdb.MutationType;
+import com.apple.foundationdb.ReadTransaction;
+import com.apple.foundationdb.Transaction;
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.provider.foundationdb.FDBDatabaseRunner;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordContext;
+import com.apple.foundationdb.record.provider.foundationdb.FDBStoreTimer;
+import com.apple.foundationdb.subspace.Subspace;
+import com.apple.foundationdb.tuple.Tuple;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+
+@API(API.Status.EXPERIMENTAL)
+public class SynchronizedSession {
+    @Nonnull
+    private Subspace lockSubspace;
+    @Nonnull
+    private FDBDatabaseRunner runner;
+
+    @Nullable
+    private UUID sessionId = null;
+
+    // TODO: configurable?
+    private static final long LEASE_PERIOD_MILL = 60_000;
+
+    private static final Object LOCK_SESSION_ID_KEY = 0L;
+    private static final Object LOCK_SESSION_TIME_KEY = 1L;
+
+    public SynchronizedSession(@Nonnull Subspace lockSubspace, @Nonnull FDBDatabaseRunner runner) {
+        this.lockSubspace = lockSubspace;
+        this.runner = runner;
+    }
+
+    // A synchronized session keeps its look by updating timestamp in each of its transaction, so do not initialize the
+    // session until it is being actively used.
+    public SynchronizedSessionRunner initializeSessionAndCreateRunner() {
```

Comment:
The semantics of this class feels kind of awkward in that the caller is expected to hand in the underlying runner, which means, I could do something like:

    syncSess = new SynchronizedSession(lockSubspace, new SynchronizedSession(otherLockSubspace,
        new FDBDatabaseRunner(database, ...).initializeSessionAndCreateRunner());

which doesn't really make any sense.  I think it would be much cleaner to, as I mentioned, combine this code into the `SynchronizedSessionRunner` and do something like:

    runner = database.newRunner(FDBStoreTimer timer, ...).toSynchronized(lockSubspace);

for starting a new session and

    runner = database.newRunner(FDBStoreTimer timer, ...).toSynchronized(lockSubspace, uuid);

for joining an existing session.